keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2262 @@
+  hard requirement is that it can connect to the API server. Optionally,
+  it may also be configured as a [special scheduler extender](#kube-scheduler).
+- *Resource kubelet plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
+
+An utility library for resource drivers will be developed outside of Kubernetes
+and does not have to be used by drivers, therefore it is not described further
+in this KEP.
+
+### State and communication
+
+A ResourceClaim object defines what kind of resource is needed and what
+the parameters for it are. It is owned by users and namespaced. Additional
+parameters are provided by a cluster admin in ResourceClass objects.
+
+The ResourceClaim spec is immutable. The ResourceClaim
+status is reserved for system usage and holds the current state of the
+resource. The status must not get lost, which in the past was not ruled
+out. For example, status could have been stored in a separate etcd instance
+with lower reliability. To recover after a loss, status was meant to be recoverable.
+A [recent KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/2527-clarify-status-observations-vs-rbac)
+clarified that status will always be stored reliably and can be used as
+proposed in this KEP.
+
+The different components communicate with each other by updating fields in the
+status. This has two advantages:
+- Changes for a resource are atomic, which avoids race conditions.
+- The only requirement for deployments is that the components can connect to
+  the API server. Direct communication is not needed, but in some cases
+  can be optionally used to improve performance.
+
+Using a single object is an intentional simplification compared to the
+PersistentVolume/PersistentVolumeClaim model for volumes where the additional
+PV object was used to capture status. That model allowed operations like
+pre-provisioning volumes and then having Kubernetes bind those to claims that
+get created later. For resources, the resource driver can and must handle such
+pre-provisioning internally. Kubernetes wouldn't know how to match
+pre-provisioned resources against claims because it has no understanding about the
+parameters.
+
+The entire state of a resource can be determined by looking at its
+ResourceClaim (see [API below](#api) for details), for example:
+
+- It is **allocated** if and only if `ResourceClaimStatus.Allocated` is non-nil and
+  points to the `AllocationResult`.
+
+- It is in use if and only if `ResourceClaimStatus.ReservedFor` contains one or
+  more users.  It does not matter whether those users, usually pods, are
+  currently running because that could change at any time.
+
+- A resource is no longer needed when it has a `DeletionTimestamp`. It must not
+  be deallocated yet when it is still in use.
+
+Some of the race conditions that need to be handled are:
+
+- A ResourceClaim gets created and deleted again while the resource driver
```

Comment:
You also have to handle the case of create/delete/create.

* Create claim "foo" for driver x
* Driver x gets an event for "foo"
* Delete "foo"
* Create claim "foo" for driver y
* Driver x adds finalizer to "foo"

The driver MUST check the UID and/or the driver when it sets the claim

Reply:
I need to call out explicitly that update operations must be used instead of patching to ensure this protection against race condition. Added:

- A ResourceClaim gets deleted and recreated while the resource driver is
  adding the finalizer. The driver must update the object to add the finalizer
  and then will get a conflict error, which informs the driver that it must
  work on a new instance of the claim. In general, patching a ResourceClaim
  is only acceptable when it does not lead to race conditions.

Reply:
Then it still doesn't cover other, incompatible changes that may have been made in the meantime. It would also need to check ResourceVersion. I'll add both here to clarify under which circumstances patching may be okay.